components/candle/progressionService.ts

```diff
@@ -0,0 +1,238 @@
+import {
+    DEFAULT_MASTERY_MAXLEVEL,
+    clampValue,
+    xpRequiredForLevel,
+    xpRequiredForEvergreenLevel,
+    levelForXp,
+    evergreenLevelForXp,
+    getMaxProfileLevel,
+} from "../utils"
+import { writeUserData } from "../databaseHandler"
+
+export class ProgressionService {
+    //NOTE: Official will always grant XP to both Location Mastery and the Player Profile
+    grantProfileProgression(
+        actionXp: number,
+        masteryXp: number,
+        drops: Unlockable[],
+        contractSession: ContractSession,
+        userProfile: UserProfile,
+    ) {
+        // Total XP for profile XP is the total sum of the action and mastery XP's
+        const xp = actionXp + masteryXp
+
+        // Grants profile XP
+        this.grantUserXp(xp, contractSession, userProfile)
+
+        // Grants Mastery Progression and Drops
+        this.grantLocationMasteryXpAndRewards(
+            masteryXp,
+            actionXp,
+            contractSession,
+            userProfile,
+        )
+
+        // Award provided drops. E.g. From challenges
+        grantDrops(userProfile.Id, drops)
+
+        // Saves profile data
+        writeUserData(userProfile.Id, contractSession.gameVersion)
+    }
+
+    // Returns current mastery progression for given location
+    getMasteryProgressionForLocation(
+        userProfile: UserProfile,
+        location: string,
+    ) {
+        return (userProfile.Extensions.progression.Locations[
+            location.toLocaleLowerCase()
+        ] ??= {
+            Xp: 0,
+            Level: 1,
+        })
+    }
+
+    // Return mastery drops from location from a level range
+    private getLocationMasteryDrops(
+        gameVersion: GameVersion,
+        isEvergreenContract: boolean,
+        masteryLocationDrops: {
+            Id: string
+            Level: number
+        }[],
+        minLevel: number,
+        maxLevel: number,
+    ) {
+        const unlockableIds = masteryLocationDrops
+            .filter((drop) => drop.Level > minLevel && drop.Level <= maxLevel)
+            .map((drop) => drop.Id)
+
+        const unlockables = getDataForUnlockables(gameVersion, unlockableIds)
+
+        /**
+         * If missions type is evergreen, checks if any of the unlockables has unlockable gear, and award those too
+         *
+         * This is required to unlock the item to the normal inventory too, as the freelancer and normal inventory item ID is not the same
+         */
+        if (isEvergreenContract) {
+            const evergreenGearUnlockables = unlockables.reduce((acc, u) => {
+                if (u.Properties.Unlocks) acc.push(...u.Properties.Unlocks)
+                return acc
+            }, [])
+            evergreenGearUnlockables.length &&
+                unlockables.push(
+                    ...getDataForUnlockables(
+                        gameVersion,
+                        evergreenGearUnlockables,
+                    ),
+                )
+        }
```

Comment:
Yeah I think this part is what went wrong. Previously the `isEvergreenContract` passed was reversed, so this wasn't executed. Now with the `isEvergreenContract` fixed, this got executed and the unlockable got added to the inventory while it shouldn't. I think the game somehow manages to add firearms unlocked from freelancer mastery to both freelancer and normal game modes, so no need to add them again here.

Reply:
The game adds them to the gear walls by checking what level you are.

Official does grant these to your inventory, but filters them from the inventory (for now at least). So that's what we should do!

Reply:
Yeah sorry, I wasnt filtering those out on the createInventory so I assumed it was working, I didnt actually realized until now that they were actually being filtered out by the pre-existing filter function, so this was only a problem when awarding them because it injects directly on the cache. Should be addressed now 👍

Reply:
We should **still** grant them, but not show them, see the CTT branch for how this was done.

Reply:
Ive added the awarding back, and filter on the function, same as createInventory. As I've mentioned tho, if the Official does have the evergreenmastery unlocks on the inventory, and we want to keep the same structure, we need to remove these filters from the inventory and filter them out at https://github.com/thepeacockproject/Peacock/blob/v6/components/menuData.ts#L360. This would keep them on the inventory as per official but not show them on the career inventory window. Let me know what you think